### DISCUSSION TITLE
About testing

### DISCUSSION THREAD

**Discussion by hyf0** [web-infra-dev/rspack#430] [General]
Discussion for testing in RSpack

@bvanjoi @IWANABETHATGUY @wre232114 @iheyunfei

**Comment by hyf0** [comment]
To be specific, we should have the [same test ability as webpack](https://github.com/webpack-contrib/mini-css-extract-plugin/tree/master/test).
I have already implemented the CSS version in https://github.com/speedy-js/rspack/tree/next/crates/rspack_plugin_css.

I think it's better than snapshots testing
- We could also test the filename
- multiple files

But I'm open to snapshots testing if it could do the same ability.

We also need to have [the same macro like swc](https://github.com/swc-project/swc/blob/b21cccfea2e9e6a44146b8b329b64e3ba79cc8b7/crates/testing_macros/tests/test.rs#L7) to [enable concurrency in tests](https://github.com/speedy-js/rspack/blob/419b2d312fe4ed3d3560a5263fd4c31066ca37b8/crates/rspack_plugin_css/tests/fixtures/mod.rs#L4-L22).

**Comment by liuliudada-w** [comment]
In my opinion, the test function have two parts. 
- Each independent function module functional module such as CSS，Assets、HTML and more should have the same test abilities as webpack( In my view this is the snapshot test )
- For each functional function, we also need unit tests. There are two main issues to discuss in this part. First, whether to force each function to have a corresponding unit test? Second, how should we control unit tests, is it written in the corresponding file or managed at the top level?

**Reply by hardfist** [reply]
we may need something like load-runner to test load behavior out of rspack implementation like webpack https://github.com/webpack/loader-runner

**Comment by hyf0** [comment]
# Conclustions
- Need the same test ability as webpack
  - Split the test ability into a single crate to reuse. #436 
  - Update snapshot automatically
  - Concurrency test macros #431 
- No unit tests for now
- No loader-runner abstraction for now
